### PR TITLE
ws: don't crash on empty branding.css

### PR DESCRIPTION
**Pull request by jelly**
cockpit_template_expand never validated if data was not `NULL` as it was used for the login page and that content was usually not empty this worked fine. With the introduction of branding support its much easier to trigger this condition.

### CODE REVIEW

File: src/ws/cockpittemplate.c [hunk 1]
```diff
@@ -92,6 +92,8 @@ cockpit_template_expand (GBytes *input,
   data = g_bytes_get_data (input, NULL);
   end = data + g_bytes_get_size (input);
 
+  g_return_val_if_fail (data != NULL, NULL);
```
Comment (allisonkarlitskaya) [inline src/ws/cockpittemplate.c]: These are not supposed to be used for conditions that can actually reach.  If you fail this then it's always a programmer error, by definition, and may abort your program depending on the environment.  You rather want to log a warning or something here and return...

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
Your commit message has a couple of typos and could use expanding to explain the exact way the issue is triggered...

File: src/ws/cockpittemplate.c (hunk 1 above)
Comment (jelly) [reply src/ws/cockpittemplate.c]: Ack!

Comment (jelly) [reply src/ws/cockpittemplate.c]: Only annoying bit is that we lack context

File: test/verify/check-connection [hunk 2]
```diff
@@ -1380,6 +1380,13 @@ until pgrep -f '^(/usr/[^ ]+/[^ /]*python[^ /]* )?/usr/bin/cockpit-bridge'; do s
             b.wait_visible("#login")
             b.assert_pixels("body", "login-screen-custom-background")
 
+            # An empty branding file does not crash cockpit-ws
+            self.allow_journal_messages("cockpit_template_expand: assertion 'data != NULL' failed")
```
Comment (Venefilyn) [inline test/verify/check-connection]: You changed the behaviour so we should look for this AFAICT

```suggestion
            self.allow_journal_messages("Empty input passed to cockpit_template_expand")
```

File: src/ws/cockpittemplate.c [hunk 3]
```diff
@@ -92,6 +92,12 @@ cockpit_template_expand (GBytes *input,
   data = g_bytes_get_data (input, NULL);
   end = data + g_bytes_get_size (input);
 
+  if (data == NULL)
+    {
+       g_warning ("Empty input passed to cockpit_template_expand");
+       return NULL;
+    }
+
```
Comment (allisonkarlitskaya) [inline src/ws/cockpittemplate.c]: This check is papering over a massive bug in find_variable().

Looking at how `data` and `end` are determined, this seems like it shouldn't be a problem.  If there is no data, `get_size()` should return `0` and `end == data` in which case obviously nothing should ever be dereferenced.

The issue is that find_variable(), instead of computing the size of the data properly using `end` is calling `strlen()`.  That desperately needs to be fixed.  I'm shocked that this has never been a problem, and I think the only reason for it is that we've never had a `branding.css` that's an exact multiple of the page size, so we've always been able to depend on the zero padding.

File: src/ws/cockpittemplate.c [hunk 4]
```diff
@@ -19,61 +19,52 @@
...
+       * c = should be '}'
        */
-      if (b != c && b + strspn (b, VARCHARS) == c)
-        break;
+      b = a + 2;
+
+      for (c = b; c < end && IS_WORDCHAR (*c); c++)
+        ;
+
+      if (b < c && c < end && *c == '}')
```
Comment (allisonkarlitskaya) [inline src/ws/cockpittemplate.c:53]: I might have added a comment here which reads "the variable name is non-empty and there is a `}` character directly following it"...

File: src/ws/cockpittemplate.c [hunk 5]
```diff
@@ -19,61 +19,52 @@
...
-      if (c == NULL)
-        return NULL;
-
-      data = c + strlen (end_marker);
-      d = data;
+  g_return_val_if_fail (before != NULL, NULL);
+  g_return_val_if_fail (after != NULL, NULL);
+  g_return_val_if_fail (data <= end, NULL);
 
+  while ((a = memmem (data, end - data, "${", 2)))
```
Comment (allisonkarlitskaya) [inline src/ws/cockpittemplate.c:36]: Maybe a comment here:

This code will return NULL (resulting in no further expansions) in case `${` is not found in the string, but will keep scanning for further potential matches in case of empty or invalid variable names (since there may be more `${` later).

This is a very slight modification of the logic in the original function (which scanned to the end of the string for `}` and then checked the string in the middle) but the result is functionally identical, except that the new version scans more or less linearly, never reading a character more than once.

That's my intent, at least.  :)

File: src/ws/test-template.c [hunk 6]
```diff
@@ -65,7 +65,7 @@ static const Fixture expand_fixtures[] = {
   { "extra-dollar-after", "Test ${oh} su${ff$ix", { "Test ", "marmalade", " su${ff$ix", NULL } },
   { "unknown", "Test ${unknown} suffix", { "Test ", "${unknown}", " suffix", NULL } },
   { "escaped", "Test \\${oh} ${oh} suffix", { "Test ", "${oh}", " ", "marmalade", " suffix", NULL } },
-  { "dash", "Test ${oh-dash} suffix", { "Test ", "dash-marmalade", " suffix", NULL } },
+  { "dash", "Test ${oh_dash} suffix", { "Test ", "dash-marmalade", " suffix", NULL } },
```
Comment (jelly) [inline src/ws/test-template.c:68]: We should also have a test for `"Test ${ig.nore} " and ideally also some unicode variables to see that they are properly ignored.

File: src/ws/cockpitbranding.c [hunk 7]
```diff
@@ -83,9 +89,9 @@ cockpit_branding_calculate_static_roots (const gchar *os_id,
   if (is_local)
     add_system_dirs (dirs);
 
-  if (os_id)
+  if (is_valid_path_component (os_id))
```
Comment (jelly) [inline src/ws/cockpitbranding.c:92]: A test for this would be nice, especially if we want to ever rewrite this :)

File: src/ws/cockpittemplate.c [hunk 8]
```diff
@@ -19,53 +19,48 @@
...
-      if (b != c && b + strspn (b, VARCHARS) == c)
-        break;
+      b = a + 2;
+
+      for (c = b; c < end && IS_WORDCHAR (*c); c++)
+        ;
+
+      if (b < c && c < end && *c == '}')
+        {
+          *before = a;
```
Comment (mvollmer) [inline src/ws/cockpittemplate.c:55]: (Just to prove I have been reading this: The old code allowed `before == NULL` and `after == NULL`, but the new code doesn't. This is now protected against by the `g_return_val_if_fail` calls above.)

File: src/ws/cockpittemplate.c (hunk 5 above)
Comment (mvollmer) [reply src/ws/cockpittemplate.c:36]: I think the new code handles `xxx${aslkhc${foobar}xyswe` differently. The old code would not find any variable, but the new code finds `${foobar}`. I think that's fine.

Comment (jelly) [reply src/ws/cockpittemplate.c:36]: This got me interested, while expanding the tests:

```
  { "var-in-string", "te${oh}st", { "temarmaladest" } },
cockpit-ws:ERROR:src/ws/test-template.c:93:test_expand: data is not the same (te != temarmaladest)
```

So that's interesting.
